### ISSUE TITLE
Digital Ocean integration

### ISSUE THREAD

**Issue by swalkinshaw** [roots/trellis#40]
Using Ansible's [digital_ocean](http://docs.ansible.com/digital_ocean_module.html) module we should be able to create a droplet and use dynamic inventory to run the playbook on it.

Would be a nice feature to also create DNS entries for a domain as an option using http://docs.ansible.com/digital_ocean_domain_module.html

- [x] I've read the [guidelines for Contributing to Roots Projects](https://github.com/roots/guidelines/blob/master/CONTRIBUTING.md)

**Comment by swalkinshaw** [comment]
Ansible still hasn't updated their DO module :(

See https://github.com/ansible/ansible-modules-core/pull/998 for a new attempt

**Comment by mAAdhaTTah** [comment]
Can we use this module now anyway? Might help them get it merged if we tested it.

**Comment by Telemakhos** [comment]
It's been merged... =)

**Comment by retlehs** [comment]
going to leave this here https://github.com/seven1m/do-install-button

**Comment by rossedman** [comment]
I think launching infrastructure and provisioning it are two different things. Terraform would be a better choice. Pass a [local-exec](https://www.terraform.io/docs/provisioners/local-exec.html) command from terraform to provision with ansible. This would allow swapping out of whole server cluster and configuration files if you wanted master-->slave databases and whatever.... My two cents.

**Comment by swalkinshaw** [comment]
@rossedman I agree that would probably be better but I'd rather keep this simple and within Ansible for now. We'll still keep it separate with an `infrastructure.yml` playbook which will only create a DO instance.

**Comment by mAAdhaTTah** [comment]
Is this in the works? Happy to test!

**Comment by swalkinshaw** [comment]
@mAAdhaTTah the DO module in Ansible is broken since DO retired V1 of their API. DO finally upgrade the module to V2 but we've been waiting on their next release to include it.

The actual work won't be difficult though. Just a simple playbook to create a droplet.

**Comment by rossedman** [comment]
@swalkinshaw I understand. I just deployed some test servers last night with Terraform and could see it working beautifully with Ansible because it keeps the concerns separated. Defining infrastructure vs Provisioning. Hopefully the DO module will update soon enough so you can do it all through Ansible.

**Comment by mAAdhaTTah** [comment]
Has there been any movement on this/are we still waiting on a new Ansible release for this to work?

**Comment by swalkinshaw** [comment]
Still waiting :sob: They haven't bumped the DO module version in forever so whenever 2.0 is released it will be included.

**Comment by swalkinshaw** [comment]
We can finally start development on this now that Ansible 2.0 is released!

**Comment by mAAdhaTTah** [comment]
The wait is over!

**Comment by davidjpeacock** [comment]
Any ideas when this will be working?  I'm still getting nowhere with Ansible 2.0.0.1

**Comment by landerss0n** [comment]
Glad to hear

**Comment by nbyloff** [comment]
Not exactly polished, but I thought I would take a swing at it. Here's a simple, working instance of creating a droplet and printing the IP address. 84221e0dfbd7275160c83fb6f27c5e6574aad46d

With the `add_host` command from the inventory module, the created droplet can immediately launch into the typical trellis server provisioning. The one thing I had to change `your_server_hostname` to `localhost ansible_connection=local` in the host file I was using. Until I did that, it would spit out an error. Don't know why.

Eventually I'd like to get it to where multiple DB, web, load balancers, etc. will all be configured. But for now, a single droplet will do. I'd like to hear what expectations on how you expect this to work and be configured. (IE: how do you want this simple demo to change in order to make it into master). This would be a great feature to have in the core.

**Comment by swalkinshaw** [comment]
Thanks for sharing @nbyloff. I've done something pretty similar to this as well. My sticking point was trying to get dynamic inventories working but thinking about it again, I'm not sure it's needed.

The typical/default setup is probably a single server so a dynamic inventory isn't really needed. It's just nice to have to skip manually editing your hosts file.

**Comment by nbyloff** [comment]
@swalkinshaw what about using a template do accomplish writing to the hosts file? The most difficult part would be to make sure you read the hosts file correctly to capture any pre-configured servers. But you could:

- Read single environment host file and collect groups & servers
- Check DO to see if the server name in inventory matches a droplet name via api
- Create droplet that doesn't exist
- add a new droplet IP address to hosts file using string replace

That might require the default hosts file to look something like this:

````
production_host ansible_host=XXX.XXX.XXX.XXX

[production]
localhost ansible_connection=local
production_host 

[web]
localhost ansible_connection=local
production_host 
````

But that will also leave the door open for updates in the future for simple configuration changes to create clusters like:

````
web1 ansible_host=XXX.XXX.XXX.XXX
web2 ansible_host=XXX.XXX.XXX.XXX
db1 ansible_host=XXX.XXX.XXX.XXX
db2 ansible_host=XXX.XXX.XXX.XXX
db3 ansible_host=XXX.XXX.XXX.XXX
lb1 ansible_host=XXX.XXX.XXX.XXX

[production]
localhost ansible_connection=local
web1

[web]
localhost ansible_connection=local
production_host 
````

**Comment by swalkinshaw** [comment]
Having a hard time picturing it, but it seems pretty brittle. Parsing an existing file and checking if names match via the DO API does not sound like fun.

**Comment by nbyloff** [comment]
Yes, while I have an idea how to keep the hosts files updated programmatically it's probably not necessary. Time could be better spent creating a playbook that builds a cluster of servers for a large website.

**Comment by davisonio** [comment]
Any update on this? Would be a cool feature. at the moment I'm doing this with terraform in the repo

**Comment by nbyloff** [comment]
Currently in my projects I modify the `wordpress_sites.yml` keys for backups like this. What if there was an optional server attribute? (Non-working psuedo code below)

    ssl:
      enabled: true
      provider: letsencrypt
    cache:
      enabled: false
    backup:
      enabled: true
      cron:
        hour: 6
        minute: 0
    server:
      enabled: true
      provider: "{{ cloud_servers.provider }}"
      name: "{{ web_server }}" # this is the name of the server; should be unique
      single: true  # single server that has DB, nginx, etc. all on one server like the current default setup

Then in like `group_vars/all/something.yml`  you could have configuration options for the droplet:

    cloud_servers:
      web_server: www-1  #unique name
      provider: digital_ocean
      image_id: ubuntu-16-04-x64
      region_id: nyc1
      ...

I added the `single: true` but might beyond the scope of this project. I mention that b/c I modified an environment and have 11 wordpress sites on 4 servers (1 LB, 2 web servers, and 1 DB server). If single were false, the options could be expanded to allow a much more dynamic environment and your users could begin building their own private clouds, especially if the community started adding playbook examples in the wiki.

A setup similar to his would allow for optional droplet creation and linking a server dynamically by a unique name vs. manually pasting the IP address in the hosts file.

Also, In the OP of this issue, @swalkinshaw mentioned updating DNS entries, which would be huge. I had mentioned a few weeks ago about Lets Encrypt and DNS validation via API on the forums. You could go end to end without any manual intervention. Exciting options!

**Comment by swalkinshaw** [comment]
Closing the oldest issue 🎉 

[trellis-cli](https://github.com/roots/trellis-cli) has DO integration in one easy command:

```bash
$ trellis droplet create
```
